### ISSUE TITLE
Jax is slower than pytorch

### ISSUE THREAD

**Issue by JiahaoYao** [google/jax#4861]
Hi, I moved the pytorch code to jax. But I find the speed is way slower, does anyone know why? I think I have `jit` added. For the runnable code, you can refer to [Colab](https://colab.research.google.com/drive/1pA65JASfS5CcS9IZDDBJgDXi4gszTgq2?usp=sharing).

The main parts doing the network optimization is here: 
```python 
 
                q_a_prime = raw_predict(params_target, s_prime)
                max_q_prime = jnp.max(q_a_prime, axis=1)
                max_q_prime = jnp.expand_dims(max_q_prime, axis=1)
                target = r + gamma * max_q_prime * done_mask

                opt_state = step(next(itercount), opt_state, s, a, target)
                params = get_params(opt_state)
```

**Comment by 8bitmp3** [comment]
Comparing PyTorch and JAX implementations side-by-side is a very good idea 👍 I should try this also.

I'd recommend checking out DM's **DQN Zoo** - https://github.com/deepmind/dqn_zoo/tree/master/dqn_zoo/dqn - for a vanilla DQN implementation with JAX and [haiku](http://github.com/deepmind/haiku) and [RLax](https://github.com/deepmind/rlax). RLax helps with RL building blocks with JAX. And haiku is like [Sonnet](https://github.com/deepmind/sonnet) (I've not used this one, to be honest).

Notice that, for example, the DM implementation in the DQN Zoo is JAX `jit`ting the gradient ascent+weight update (I think) _and_ the action sampling for the RL agent https://github.com/deepmind/dqn_zoo/blob/master/dqn_zoo/dqn/agent.py:

```python
...
    def update(rng_key, opt_state, online_params, target_params, transitions):
      """Computes learning update from batch of replay transitions."""
      rng_key, update_key = jax.random.split(rng_key)
      d_loss_d_params = jax.grad(loss_fn)(online_params, target_params,
                                          transitions, update_key)
      updates, new_opt_state = optimizer.update(d_loss_d_params, opt_state)
      new_online_params = optax.apply_updates(online_params, updates)
      return rng_key, new_opt_state, new_online_params

    self._update = jax.jit(update)

    def select_action(rng_key, network_params, s_t, exploration_epsilon):
      """Samples action from eps-greedy policy wrt Q-values at given state."""
      rng_key, apply_key, policy_key = jax.random.split(rng_key, 3)
      q_t = network.apply(network_params, apply_key, s_t[None, ...]).q_values[0]
      a_t = rlax.epsilon_greedy().sample(policy_key, q_t, exploration_epsilon)
      return rng_key, a_t

    self._select_action = jax.jit(select_action)
...
```

Also, maybe you could try using **Flax** ([google/flax](https://github.com/google/flax)) instead of Stax and then check if there're any possible improvements - WDYT? There's probably other stuff like what's happening in PyTorchT (C++) vs JAX (mostly Python) under the hood - that's a forever mystery to me. 🤔

cc @dynamicwebpaige (DM UX 👋 )

P.S. Deep RL experiments are super challenging to reproduce, as you probably know — the design choices matter (The Mirage of Action-Dependent Baselines in Reinforcement Learning by Tucker et al, 2018 https://arxiv.org/pdf/1802.10031.pdf) and there's sensitivity to hyperparams, random seeds/trials, the reward scale, and network architecture (Deep Reinforcement Learning that Matters by Henderson et al, 2019 https://arxiv.org/pdf/1709.06560.pdf).) Check out this library — SLM-Lab — built with PyTorch, it's pretty awesome. The authors also have a cool book - I highly recommend it: https://slm-lab.gitbook.io/slm-lab/publications-and-talks/instruction-for-the-book-+-intro-to-rl-section (cc @lgraesser @kengz)

**Comment by JiahaoYao** [comment]
Cool @8bitmp3! Awesome feedback! I will try Flax and probably get back to you. Also, thanks for sharing a lot of information! That is indeed helpful!

**Comment by froystig** [comment]
It seems like there's no action for us to take here, so I'm going to close this issue. Please re-open or file something new if there's a particular performance issue you encounter.

(Thanks @8bitmp3!)
